parser/src/aspects/match.rs

```diff
@@ -0,0 +1,526 @@
+    fn parse_match<P>(&mut self, parse_arm: P) -> ParseResult<Match<'a>>
+        where P: FnMut(&mut Self) -> ParseResult<Expr<'a>> + Clone;
+}
+
+impl<'a> MatchAspect<'a> for Parser<'a> {
+    fn parse_match<P>(&mut self, parse_arm: P) -> ParseResult<Match<'a>>
+        where P: FnMut(&mut Self) -> ParseResult<Expr<'a>> + Clone {
+        self.cursor.force(
+            of_type(TokenType::Match).and_then(blanks()),
+            "expected 'match' keyword at start of match expression.",
+        )?;
+
+        let operand = Box::new(self.expression_statement()?);
+
+        let arms = self.parse_match_arms(parse_arm)?;
+
+        Ok(Match {
+            operand,
+            arms
+        })
+    }
+}
+
+impl<'a> Parser<'a> {
+    fn parse_match_arms<P>(&mut self, parse_arm: P) -> ParseResult<Vec<MatchArm<'a>>>
+        where P: FnMut(&mut Self) -> ParseResult<Expr<'a>> + Clone {
+        self.cursor.force(blanks().then(of_type(CurlyLeftBracket)), "expected '{'")?;
+
+        let mut arms: Vec<MatchArm<'a>> = Vec::new();
+
+        while self.cursor.advance(blanks().then(of_type(CurlyRightBracket))).is_none() {
+            let arm = self.parse_match_arm(parse_arm.clone())?;
+            arms.push(arm);
+        }
+
+        Ok(arms)
+    }
+
+    fn parse_match_arm<P>(&mut self, parse_arm: P) -> ParseResult<MatchArm<'a>>
+        where P: FnMut(&mut Self) -> ParseResult<Expr<'a>> {
+        self.cursor.advance(blanks()); //consume blanks
+
+        let val_name = self.parse_extracted_name()?;
+        let patterns = self.parse_patterns()?;
+        let guard = self.parse_guard()?;
+        let body = Box::new(self.parse_body(parse_arm)?);
+
+        Ok(MatchArm {
+            val_name,
+            patterns,
+            guard,
+            body,
+        })
+    }
+
+    fn parse_extracted_name(&mut self) -> ParseResult<Option<&'a str>> {
+        if self.cursor.lookahead(
+            aerated(any()).then(of_type(At))
+        ).is_some() {
+            let name = self.cursor.force(of_type(Identifier), "expected identifier for extracted val name")
+                .map(|t| t.value)
+                .map(Some);
+            //consume everything before 'at' included
+            self.cursor.advance(repeat(not(of_type(At)).and_then(any())).then(any()));
+            name
+        } else {
+            Ok(None)
+        }
+    }
+
+    fn parse_patterns(&mut self) -> ParseResult<Vec<MatchPattern<'a>>> {
+        macro_rules! is_at_pattern_end {
+            () =>{ self.cursor.lookahead(blanks().then(of_types(&[If, FatArrow]))).is_some() }
+        }
+
+        if is_at_pattern_end!() {
+            return self.expected("required pattern")
+        }
+
+        let first = self.parse_pattern()?;
+
+        if first == Wildcard {
+            return if !is_at_pattern_end!() {
+                self.expected("wildcard pattern cannot be followed by other patterns")
+            } else {
+                Ok(vec![first])
+            }
```

Comment:
```suggestion
            return if is_at_pattern_end!() {
                Ok(vec![first])
            } else {
                self.expected("wildcard pattern cannot be followed by other patterns")
            }
```
Don't make unnecessary condition negations.